apps/core/views/tests.py

```diff
@@ -223,10 +223,10 @@ def __get_history(self, period_ids):
             try:
                 period_id = TaskPeriodSchedule.objects.filter(
                     id__in=period_ids, date_create__gt=change.date).order_by("counter")[0].id
-                if period_id not in history[change.test.id]:
-                    history[change.test.id][period_id] = list()
                 if change.test.id not in history:
                     history[change.test.id] = dict()
+                if period_id not in history[change.test.id]:
```

Comment:
there is a built-in method for dictionary type for adding a key with default value in case a key does not exist:
[setdefault](https://docs.python.org/2/library/stdtypes.html#dict.setdefault)

```python
history[change.test.id].setdefault(period_id, dict())
```
..should help you saving few lines of code and keeping away from conditional hell :)